MovingBlocks/Terasology
# feat: Reviving of "UI for editing arbitrary types and AutoConfigs"

### Contains 

TypeWidgetLibrary: A system to generate a UIWidget bound to an object/Binding of any arbitrary type.
NUI support for FlexibleConfig (will potentially be repurposed to support AutoConfigs introduced in #3723)

Updated, Merged with develop and splitted in TeraNUI too.

TeraNUI's PR: https://github.com/MovingBlocks/TeraNUI/pull/36
Should work together.

### How to test(currently test screens)

1. open game
2. go to "settings"
3.a. check `autoconfig` screen
3.b. check `builtin` screen ( type command `showScreen builtinTypeWidgetTestScreen`)
3.c. check `math` screen  ( type command `showScreen mathTypeWidgetTestScreen`)



### Outstanding

* [x] Autoconfig translate
* [x] Autoconfig's widgets buildings via `TypeWidgetLibrary`
* [x] Autoconfig's ui documentation.
* [x] Remove test screens (using for demonstration purpose)
* [ ] replace AudioSettingScreen with AutoConfigScreen (AudioConfig already uses AutoConfig)


### Follow-up (don't see there while don't merged)

* Split AutoConfigs at loading scopes (add feature): 
   * GLOBAL - configs for entire game(all stages/subsystems, etc). Provides by subsystems, engine itself.
   * GAME - configs for entire game, with all worlds. Provides by subsystems, engine, modules.
   * WORLD - configs for world.  Provides by subsystems, engine, modules. (ideally for world generator's configs)
 >GAME and WORLD configs should transfer via network in multiplayer(if needs)

* Split AutoConfig's Settings by "user roles"(codename):
  * usual player - can be changed in Settings menu
  * advanced player - can be changed in Settings -> Advanced Settings menu
  * internal/no gui - cannot be changed by GUI (needs or special mechanic, or code only usage)

* Try integrate TypeWidgetLibrary in some exists screens.

### Link to previous PR:

close #3699 
Copy from https://github.com/MovingBlocks/Terasology/pull/3699 :


This PR has two major contents:

    TypeWidgetLibrary: A system to generate a UIWidget bound to an object/Binding of any arbitrary type.
    NUI support for FlexibleConfig (will potentially be repurposed to support AutoConfigs introduced in #3723)

Arbitrary Type Widgets

Used to generate UIWidgets to edit any arbitrary type. Follows a system of generating widgets through TypeWidgetFactorys stored in a TypeWidgetLibrary, very similar to the TypeHandlerLibrary-TypeHandlerFactory system introduced in #3535.

You can test the widgets in the builtinTypeWidgetTestScreen (widgets for types like collections, arrays, numbers, strings, arbitrary objects) and the mathTypeWidgetTestScreen (TeraMath types like vectors, rects, and quaternions).
TypeWidgetLibrary

Contains a number of TypeWidgetFactorys which are queried to create UIWidgets. The UIWidget for an object is obtained by querying each TypeWidgetFactory until one makes a UIWidget successfully.
TypeWidgetFactory

Creates UIWidgets for objects of a specific structure/type.
TODO

    Documentation

Nice-to-haves

    Support recursive types by separating widget creation and binding process and caching created widgets
        Dependency: Introduce widget cloning to clone cached widgets (maybe serialize-deserialize via #3535?)

FlexibleConfig/AutoConfig UI

Introduces a prototype for how UI for FlexibleConfigs will work. This PR will ultimately contain a fully working NUI frontend for FlexibleConfigs.
Workflow

TODO: Obsolete, needs to be updated

The main components of the FC NUI frontend are FlexibleConfigWidgetFactory, SettingWidgetFactory, and SettingWidget. For a usage example, please look at FlexibleConfigTestScreen. The screen can be loaded in the menus using showScreen flexibleConfigTest; it need not be loaded in-game.
FlexibleConfigWidgetFactory

The FlexibleConfigWidgetFactory is used to build widgets that render entire FCs, i.e. render all their settings. Since settings will never be used in isolation, all a caller has to do to render UI for a FlexibleConfig is call the FlexibleConfigWidgetFactory.buildWidgetFor(FlexibleConfig) method, which returns a UIWidget that can be used anywhere.

The FlexibleConfigWidgetFactory requires a ModuleManager and an AssetManager -- the former for getting the ModuleEnvironment to load SettingWidgets from, and the latter to load widgets stored in a .ui asset.
buildWidgetFor

the buildWidgetFor method builds a PropertyLayout with its settings as properties. The widget generation for a specific setting is delegated to SettingWidgetFactory.
SettingWidgetFactory

SettingWidgetFactory is responsible for creating, via reflection, a SettingWidget bound to a specific Setting. Since the constraint on a setting is truly what determines how its UI must be rendered (i.e. what options to give the user), it uses the type of SettingConstraint on the setting to load the corresponding SettingWidget. Since settings will never be used in isolation, this class will never have to be used elsewhere and can (should) safely be made package-private.
SettingWidget

The primary responsibility to render UI for a Setting falls onto the SettingWidget. It is an abstract class generic over the type of value and constraint of the setting it is bound to. Much like screens, the template for a SettingWidget can specified in a .ui file, whose URI must be specified in the SettingWidget constructor.
initialise

The widget specified in the .ui file is loaded (via loadContents) and the SettingWidget is bound to a Setting (retrievable via getSetting) before the abstract method initialise is called. The initialise method is meant to allow subclasses to add listeners and Bindings to UI widgets to set the value in the setting. This method is implemented much like CoreScreenLayer.initialise; widgets can be found using the find method and can be set up to modify the value in the bound Setting.
TODO

Documentation
Support no (null) constraint via TypeWidgetLibrary
Translation support for FlexibleConfig and Setting names and descriptions
Package protect internal classes (like SettingWidgetFactory)
